File: AmericanExpress/src/test/java/com/braintreepayments/api/americanexpress/AmericanExpressClientUnitTest.java
package com.braintreepayments.api.americanexpress;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertNotNull;
import static junit.framework.Assert.assertNull;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

import com.braintreepayments.api.core.AnalyticsEventParams;
import com.braintreepayments.api.testutils.Fixtures;
import com.braintreepayments.api.testutils.MockBraintreeClientBuilder;
import com.braintreepayments.api.core.BraintreeClient;
import com.braintreepayments.api.sharedutils.AuthorizationException;
import com.braintreepayments.api.sharedutils.HttpResponseCallback;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.robolectric.RobolectricTestRunner;

@RunWith(RobolectricTestRunner.class)
public class AmericanExpressClientUnitTest {

    private AmericanExpressGetRewardsBalanceCallback amexRewardsCallback;

    @Before
    public void beforeEach() {
        amexRewardsCallback = mock(AmericanExpressGetRewardsBalanceCallback.class);
    }

    @Test
    public void getRewardsBalance_sendsGETRequestForAmexAwardsBalance() {
        BraintreeClient braintreeClient = mock(BraintreeClient.class);
        AmericanExpressClient sut = new AmericanExpressClient(braintreeClient);
        sut.getRewardsBalance("fake-nonce", "USD", amexRewardsCallback);

        ArgumentCaptor<String> urlCaptor = ArgumentCaptor.forClass(String.class);
        verify(braintreeClient).sendGET(urlCaptor.capture(), any(HttpResponseCallback.class));

        String url = urlCaptor.getValue();
        assertEquals(
                "/v1/payment_methods/amex_rewards_balance?paymentMethodNonce=fake-nonce&currencyIsoCode=USD",
                url);
    }

    @Test
    public void getRewardsBalance_callsListenerWithRewardsBalanceOnSuccess() {
        BraintreeClient braintreeClient = new MockBraintreeClientBuilder()
                .sendGETSuccessfulResponse(Fixtures.AMEX_REWARDS_BALANCE_SUCCESS)
                .build();

        AmericanExpressClient sut = new AmericanExpressClient(braintreeClient);
        sut.getRewardsBalance("fake-nonce", "USD", amexRewardsCallback);

        ArgumentCaptor<AmericanExpressResult> amexRewardsCaptor =
                ArgumentCaptor.forClass(AmericanExpressResult.class);
        verify(amexRewardsCallback).onAmericanExpressResult(amexRewardsCaptor.capture());

        AmericanExpressResult result = amexRewardsCaptor.getValue();
        assertTrue(result instanceof AmericanExpressResult.Success);
        AmericanExpressRewardsBalance rewardsBalance = ((AmericanExpressResult.Success) result).getRewardsBalance();
        assertNotNull(rewardsBalance);
        assertEquals("0.0070", rewardsBalance.getConversionRate());
        assertEquals("316795.03", rewardsBalance.getCurrencyAmount());
        assertEquals("USD", rewardsBalance.getCurrencyIsoCode());
        assertEquals("715f4712-8690-49ed-8cc5-d7fb1c2d", rewardsBalance.getRequestId());
        assertEquals("45256433", rewardsBalance.getRewardsAmount());
        assertEquals("Points", rewardsBalance.getRewardsUnit());
        assertNull(rewardsBalance.getErrorCode());
        assertNull(rewardsBalance.getErrorMessage());
    }

    @Test
    public void getRewardsBalance_callsListenerWithRewardsBalanceWithErrorCode_OnIneligibleCard() {
        BraintreeClient braintreeClient = new MockBraintreeClientBuilder()
                .sendGETSuccessfulResponse(Fixtures.AMEX_REWARDS_BALANCE_INELIGIBLE_CARD)
                .build();

        AmericanExpressClient sut = new AmericanExpressClient(braintreeClient);
        sut.getRewardsBalance("fake-nonce", "USD", amexRewardsCallback);

        ArgumentCaptor<AmericanExpressResult> amexRewardsCaptor =
                ArgumentCaptor.forClass(AmericanExpressResult.class);
        verify(amexRewardsCallback).onAmericanExpressResult(amexRewardsCaptor.capture());

        AmericanExpressResult result = amexRewardsCaptor.getValue();
        assertTrue(result instanceof AmericanExpressResult.Success);
        AmericanExpressRewardsBalance rewardsBalance = ((AmericanExpressResult.Success) result).getRewardsBalance();
        assertNotNull(rewardsBalance);
        assertNull(rewardsBalance.getConversionRate());
        assertNull(rewardsBalance.getCurrencyAmount());
        assertNull(rewardsBalance.getCurrencyIsoCode());
        assertNull(rewardsBalance.getRequestId());
        assertNull(rewardsBalance.getRewardsAmount());
        assertNull(rewardsBalance.getRewardsUnit());
        assertEquals("INQ2002", rewardsBalance.getErrorCode());
        assertEquals("Card is ineligible", rewardsBalance.getErrorMessage());
    }

    @Test
    public void getRewardsBalance_callsListenerWithRewardsBalanceWithErrorCode_OnInsufficientPoints() {
        BraintreeClient braintreeClient = new MockBraintreeClientBuilder()
                .sendGETSuccessfulResponse(Fixtures.AMEX_REWARDS_BALANCE_INSUFFICIENT_POINTS)
                .build();

        AmericanExpressClient sut = new AmericanExpressClient(braintreeClient);
        sut.getRewardsBalance("fake-nonce", "USD", amexRewardsCallback);

        ArgumentCaptor<AmericanExpressResult> amexRewardsCaptor =
                ArgumentCaptor.forClass(AmericanExpressResult.class);
        verify(amexRewardsCallback).onAmericanExpressResult(amexRewardsCaptor.capture());

        AmericanExpressResult result = amexRewardsCaptor.getValue();
        assertTrue(result instanceof AmericanExpressResult.Success);

        AmericanExpressRewardsBalance rewardsBalance = ((AmericanExpressResult.Success) result).getRewardsBalance();
        assertNotNull(rewardsBalance);
        assertNull(rewardsBalance.getConversionRate());
        assertNull(rewardsBalance.getCurrencyAmount());
        assertNull(rewardsBalance.getCurrencyIsoCode());
        assertNull(rewardsBalance.getRequestId());
        assertNull(rewardsBalance.getRewardsAmount());
        assertNull(rewardsBalance.getRewardsUnit());
        assertEquals("INQ2003", rewardsBalance.getErrorCode());
        assertEquals("Insufficient points on card", rewardsBalance.getErrorMessage());
    }

    @Test
    public void getRewardsBalance_callsBackFailure_OnHttpError() {
        Exception expectedError = new Exception("error");
        BraintreeClient braintreeClient = new MockBraintreeClientBuilder()
                .sendGETErrorResponse(expectedError)
                .build();

        AmericanExpressClient sut = new AmericanExpressClient(braintreeClient);
        sut.getRewardsBalance("fake-nonce", "USD", amexRewardsCallback);

        ArgumentCaptor<AmericanExpressResult> amexRewardsCaptor =
                ArgumentCaptor.forClass(AmericanExpressResult.class);
        verify(amexRewardsCallback).onAmericanExpressResult(amexRewardsCaptor.capture());

        AmericanExpressResult result = amexRewardsCaptor.getValue();
        assertTrue(result instanceof AmericanExpressResult.Failure);
        Exception actualError = ((AmericanExpressResult.Failure) result).getError();
        assertEquals(expectedError, actualError);
    }


    @Test
    public void getRewardsBalance_sendsAnalyticsEventOnSuccess() {
        BraintreeClient braintreeClient = new MockBraintreeClientBuilder()
                .sendGETSuccessfulResponse(Fixtures.AMEX_REWARDS_BALANCE_SUCCESS)
                .build();

        AmericanExpressClient sut = new AmericanExpressClient(braintreeClient);
        sut.getRewardsBalance("fake-nonce", "USD", amexRewardsCallback);

        AnalyticsEventParams params = new AnalyticsEventParams();
        verify(braintreeClient).sendAnalyticsEvent(AmericanExpressAnalytics.REWARDS_BALANCE_STARTED, params, true);
        verify(braintreeClient).sendAnalyticsEvent(AmericanExpressAnalytics.REWARDS_BALANCE_SUCCEEDED, params, true);
    }

    @Test
    public void getRewardsBalance_sendsAnalyticsEventOnFailure() {
        BraintreeClient braintreeClient = new MockBraintreeClientBuilder()
                .sendGETErrorResponse(new AuthorizationException("Bad fingerprint"))
                .build();

        AmericanExpressClient sut = new AmericanExpressClient(braintreeClient);
        sut.getRewardsBalance("fake-nonce", "USD", amexRewardsCallback);

        AnalyticsEventParams params = new AnalyticsEventParams();
        AnalyticsEventParams errorParams = new AnalyticsEventParams(null, false, null, null, null, null, null, null, null, null, null, "Bad fingerprint");
        verify(braintreeClient).sendAnalyticsEvent(AmericanExpressAnalytics.REWARDS_BALANCE_STARTED, params, true);
        verify(braintreeClient).sendAnalyticsEvent(AmericanExpressAnalytics.REWARDS_BALANCE_FAILED, errorParams, true);
    }

    @Test
    public void getRewardsBalance_sendsAnalyticsEventOnParseError() {
        String notJson = "Big blob that is not a valid JSON object";
        BraintreeClient braintreeClient = new MockBraintreeClientBuilder()
                .sendGETSuccessfulResponse(notJson)
                .build();
        AmericanExpressClient sut = new AmericanExpressClient(braintreeClient);
        sut.getRewardsBalance("fake-nonce", "USD", amexRewardsCallback);

        AnalyticsEventParams params = new AnalyticsEventParams();
        AnalyticsEventParams errorParams = new AnalyticsEventParams(null, false, null, null, null, null, null, null, null, null, null, "Value " + notJson.split(" ")[0] + " of type java.lang.String cannot be converted to JSONObject");
        verify(braintreeClient).sendAnalyticsEvent(AmericanExpressAnalytics.REWARDS_BALANCE_STARTED, params, true);
        verify(braintreeClient).sendAnalyticsEvent(AmericanExpressAnalytics.REWARDS_BALANCE_FAILED, errorParams, true);
    }
}

File: AmericanExpress/src/test/java/com/braintreepayments/api/americanexpress/AmericanExpressClientUnitTest.kt
package com.braintreepayments.api.americanexpress

import com.braintreepayments.api.core.AnalyticsEventParams
import com.braintreepayments.api.core.BraintreeClient
import com.braintreepayments.api.sharedutils.AuthorizationException
import com.braintreepayments.api.testutils.Fixtures
import com.braintreepayments.api.testutils.MockkBraintreeClientBuilder
import io.mockk.coEvery
import io.mockk.coVerify
import io.mockk.mockk
import io.mockk.slot
import io.mockk.verify
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.StandardTestDispatcher
import kotlinx.coroutines.test.TestScope
import kotlinx.coroutines.test.advanceUntilIdle
import kotlinx.coroutines.test.runTest
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import java.io.IOException
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

@OptIn(ExperimentalCoroutinesApi::class)
@RunWith(RobolectricTestRunner::class)
class AmericanExpressClientUnitTest {

    private lateinit var amexRewardsCallback: AmericanExpressGetRewardsBalanceCallback

    @Before
    fun beforeEach() {
        amexRewardsCallback = mockk(relaxed = true)
    }

    @Test
    fun `when getRewardsBalance is called, GET request includes nonce and currency code in query params`() = runTest {
        val braintreeClient = mockk<BraintreeClient>(relaxed = true)
        coEvery { braintreeClient.sendGET(any()) } returns """
        {
            "conversionRate": "0.0070",
            "currencyAmount": "316795.03",
            "currencyIsoCode": "USD",
            "requestId": "request-id",
            "rewardsAmount": "45256433",
            "rewardsUnit": "Points"
        }
        """.trimIndent()
        val testDispatcher = StandardTestDispatcher(testScheduler)
        val testScope = TestScope(testDispatcher)
        val sut = AmericanExpressClient(braintreeClient, testDispatcher, testScope)

        sut.getRewardsBalance("fake-nonce", "USD", amexRewardsCallback)
        advanceUntilIdle()

        val urlSlot = slot<String>()
        coVerify { braintreeClient.sendGET(capture(urlSlot)) }

        val url = urlSlot.captured
        assertEquals(
            "/v1/payment_methods/amex_rewards_balance?paymentMethodNonce=fake-nonce&currencyIsoCode=USD",
            url
        )
    }

    @Test
    fun `when getRewardsBalance receives a successful response, callback receives success result with rewards balance`() = runTest {
        val braintreeClient = MockkBraintreeClientBuilder().sendGetSuccessfulResponse(
            Fixtures.AMEX_REWARDS_BALANCE_SUCCESS).build()

        val testDispatcher = StandardTestDispatcher(testScheduler)
        val testScope = TestScope(testDispatcher)
        val sut = AmericanExpressClient(braintreeClient, testDispatcher, testScope)
        sut.getRewardsBalance("fake-nonce", "USD", amexRewardsCallback)
        advanceUntilIdle()
        val amexRewardsSlot = slot<AmericanExpressResult>()
        verify { amexRewardsCallback.onAmericanExpressResult(capture(amexRewardsSlot)) }

        val result = amexRewardsSlot.captured
        assertTrue(result is AmericanExpressResult.Success)

        val rewardsBalance = result.rewardsBalance
        assertNotNull(rewardsBalance)
        assertEquals("0.0070", rewardsBalance.conversionRate)
        assertEquals("316795.03", rewardsBalance.currencyAmount)
        assertEquals("USD", rewardsBalance.currencyIsoCode)
        assertEquals("715f4712-8690-49ed-8cc5-d7fb1c2d", rewardsBalance.requestId)
        assertEquals("45256433", rewardsBalance.rewardsAmount)
        assertEquals("Points", rewardsBalance.rewardsUnit)
        assertNull(rewardsBalance.errorCode)
        assertNull(rewardsBalance.errorMessage)
    }

    @Test
    fun `when getRewardsBalance is called and card is ineligible, callback receives success result with error code`() = runTest {
        val braintreeClient = MockkBraintreeClientBuilder().sendGetSuccessfulResponse(
            Fixtures.AMEX_REWARDS_BALANCE_INELIGIBLE_CARD).build()
        val testDispatcher = StandardTestDispatcher(testScheduler)
        val testScope = TestScope(testDispatcher)
        val sut = AmericanExpressClient(braintreeClient, testDispatcher, testScope)
        sut.getRewardsBalance("fake-nonce", "USD", amexRewardsCallback)
        advanceUntilIdle()
        val amexRewardsSlot = slot<AmericanExpressResult>()
        verify { amexRewardsCallback.onAmericanExpressResult(capture(amexRewardsSlot)) }

        val result = amexRewardsSlot.captured
        assertTrue(result is AmericanExpressResult.Success)

        val rewardsBalance = result.rewardsBalance
        assertNotNull(rewardsBalance)
        assertNull(rewardsBalance.conversionRate)
        assertNull(rewardsBalance.currencyAmount)
        assertNull(rewardsBalance.currencyIsoCode)
        assertNull(rewardsBalance.requestId)
        assertNull(rewardsBalance.rewardsAmount)
        assertNull(rewardsBalance.rewardsUnit)
        assertEquals("INQ2002", rewardsBalance.errorCode)
        assertEquals("Card is ineligible", rewardsBalance.errorMessage)
    }

    @Test
    fun `when getRewardsBalance is called and points are insufficient, callback receives success result with error code`() = runTest {
        val braintreeClient = MockkBraintreeClientBuilder().sendGetSuccessfulResponse(
            Fixtures.AMEX_REWARDS_BALANCE_INSUFFICIENT_POINTS).build()
        val testDispatcher = StandardTestDispatcher(testScheduler)
        val testScope = TestScope(testDispatcher)
        val sut = AmericanExpressClient(braintreeClient, testDispatcher, testScope)
        sut.getRewardsBalance("fake-nonce", "USD", amexRewardsCallback)
        advanceUntilIdle()
        val amexRewardsSlot = slot<AmericanExpressResult>()
        verify { amexRewardsCallback.onAmericanExpressResult(capture(amexRewardsSlot)) }

        val result = amexRewardsSlot.captured
        assertTrue(result is AmericanExpressResult.Success)

        val rewardsBalance = result.rewardsBalance
        assertNotNull(rewardsBalance)
        assertNull(rewardsBalance.conversionRate)
        assertNull(rewardsBalance.currencyAmount)
        assertNull(rewardsBalance.currencyIsoCode)
        assertNull(rewardsBalance.requestId)
        assertNull(rewardsBalance.rewardsAmount)
        assertNull(rewardsBalance.rewardsUnit)
        assertEquals("INQ2003", rewardsBalance.errorCode)
        assertEquals("Insufficient points on card", rewardsBalance.errorMessage)
    }

    @Test
    fun `when getRewardsBalance http request errors, callback receives failure result`() = runTest {
        val expectedError = IOException("error")
        val braintreeClient = MockkBraintreeClientBuilder().sendGetErrorResponse(
            expectedError).build()
        val testDispatcher = StandardTestDispatcher(testScheduler)
        val testScope = TestScope(testDispatcher)
        val sut = AmericanExpressClient(braintreeClient, testDispatcher, testScope)
        sut.getRewardsBalance("fake-nonce", "USD", amexRewardsCallback)
        advanceUntilIdle()
        val amexRewardsSlot = slot<AmericanExpressResult>()
        verify { amexRewardsCallback.onAmericanExpressResult(capture(amexRewardsSlot)) }

        val result = amexRewardsSlot.captured
        assertTrue(result is AmericanExpressResult.Failure)

        val actualError = result.error
        assertEquals(expectedError, actualError)
    }

    @Test
    fun `when getRewardsBalance succeeds, started and succeeded analytics events are sent`() = runTest {
        val braintreeClient = MockkBraintreeClientBuilder().sendGetSuccessfulResponse(
            Fixtures.AMEX_REWARDS_BALANCE_SUCCESS).build()
        val testDispatcher = StandardTestDispatcher(testScheduler)
        val testScope = TestScope(testDispatcher)
        val sut = AmericanExpressClient(braintreeClient, testDispatcher, testScope)
        sut.getRewardsBalance("fake-nonce", "USD", amexRewardsCallback)
        advanceUntilIdle()
        val params = AnalyticsEventParams()
        verify { braintreeClient.sendAnalyticsEvent(AmericanExpressAnalytics.REWARDS_BALANCE_STARTED, params, true) }
        verify {
            braintreeClient.sendAnalyticsEvent(AmericanExpressAnalytics.REWARDS_BALANCE_SUCCEEDED, params, true)
        }
    }

    @Test
    fun `when getRewardsBalance http request fails, started and failed analytics events are sent with error description`() = runTest {
        val braintreeClient = MockkBraintreeClientBuilder().sendGetErrorResponse(
            AuthorizationException("Bad fingerprint")).build()
        val testDispatcher = StandardTestDispatcher(testScheduler)
        val testScope = TestScope(testDispatcher)
        val sut = AmericanExpressClient(braintreeClient, testDispatcher, testScope)
        sut.getRewardsBalance("fake-nonce", "USD", amexRewardsCallback)
        advanceUntilIdle()
        val params = AnalyticsEventParams()
        val errorParams = AnalyticsEventParams(errorDescription = "Bad fingerprint")

        verify { braintreeClient.sendAnalyticsEvent(AmericanExpressAnalytics.REWARDS_BALANCE_STARTED, params, true) }
        verify {
            braintreeClient.sendAnalyticsEvent(AmericanExpressAnalytics.REWARDS_BALANCE_FAILED, errorParams, true)
        }
    }

    @Test
    fun `when getRewardsBalance response has a json parse error, started and failed analytics events are sent with error description`() = runTest {
        val notJson = "Big blob that is not a valid JSON object"
        val braintreeClient = MockkBraintreeClientBuilder().sendGetSuccessfulResponse(notJson).build()
        val testDispatcher = StandardTestDispatcher(testScheduler)
        val testScope = TestScope(testDispatcher)
        val sut = AmericanExpressClient(braintreeClient, testDispatcher, testScope)
        sut.getRewardsBalance("fake-nonce", "USD", amexRewardsCallback)
        advanceUntilIdle()
        val params = AnalyticsEventParams()
        val errorParams = AnalyticsEventParams(
            errorDescription =
                "Value ${notJson.split(" ")[0]} of type java.lang.String cannot be converted to JSONObject"
        )

        verify { braintreeClient.sendAnalyticsEvent(AmericanExpressAnalytics.REWARDS_BALANCE_STARTED, params, true) }
        verify {
            braintreeClient.sendAnalyticsEvent(AmericanExpressAnalytics.REWARDS_BALANCE_FAILED, errorParams, true)
        }
    }
}
